lib/src/outputs/advanced_file_output.dart

```diff
@@ -0,0 +1,147 @@
+              Level.fatal,
+              Level.warning,
+              // ignore: deprecated_member_use_from_same_package
+              Level.wtf,
+            ],
+        assert(
+          (file != null ? 1 : 0) + (directory != null ? 1 : 0) == 1,
+          'Either file or directory must be set',
+        );
+
+  final File? file;
+  final Directory? directory;
+
+  final bool overrideExisting;
+  final Encoding encoding;
+
+  final List<Level> writeImmediately;
+  final Duration maxDelay;
+  final int maxLogFileSizeMB;
+  final int maxBufferSize;
+
+  IOSink? _sink;
+  File? _targetFile;
+  Timer? _bufferWriteTimer;
+  Timer? _targetFileUpdater;
+
+  final List<OutputEvent> _buffer = [];
+
+  bool get dynamicFilesMode => directory != null;
+  File? get targetFile => _targetFile;
+
+  @override
+  Future<void> init() async {
+    if (dynamicFilesMode) {
+      //we use sync directory check to avoid losing
+      //potential initial boot logs in early crash scenarios
+      if (!directory!.existsSync()) {
+        directory!.createSync(recursive: true);
+      }
+
+      _targetFileUpdater = Timer.periodic(
+        const Duration(minutes: 1),
+        (_) => _updateTargetFile(),
+      );
+    }
+
+    _bufferWriteTimer = Timer.periodic(maxDelay, (_) => _writeOutBuffer());
+    await _updateTargetFile(); //run first setup without waiting for timer tick
+  }
+
+  @override
+  void output(OutputEvent event) {
+    _buffer.add(event);
+    // If event level is present in writeImmediately, write out the buffer
+    // along with any other possible elements that accumulated in it since
+    // the last timer tick
+    // Also write out if buffer is overfilled
+    if (_buffer.length > maxBufferSize ||
+        writeImmediately.contains(event.level)) {
+      _writeOutBuffer();
+    }
+  }
+
+  void _writeOutBuffer() {
+    if (_sink == null) return; //wait until _sink becomes available
+    for (final event in _buffer) {
+      _sink?.writeAll(event.lines, Platform.lineTerminator);
+      _sink?.writeln();
+    }
+    _buffer.clear();
+  }
+
+  Future<void> _updateTargetFile() async {
+    if (!dynamicFilesMode) {
+      await _openFile(file!);
+      return;
+    }
+
+    final t = DateTime.now();
+    final newName =
+        '${t.year}-${t.month.twoDigits}-${t.day.twoDigits}_${t.hour.twoDigits}-${t.minute.twoDigits}-${t.second.twoDigits}-${t.millisecond.threeDigits}';
+    if (_targetFile == null) {
+      // just create a new file on first boot
+      await _openFile(File('${directory!.path}/${newName}_init.txt'));
+    } else {
+      final proposed = File('${directory!.path}/${newName}_next.txt');
+      try {
+        if (await _targetFile!.length() > maxLogFileSizeMB * 1000000) {
+          await _closeCurrentFile();
+          await _openFile(proposed);
+        }
+      } catch (e) {
+        // try creating another file and working with it
+        await _closeCurrentFile();
+        await _openFile(proposed);
+      }
```

Comment:
I don't think so, because _openFile(proposed) is not meant to be called if current file size is smaller than the limit.

Reply:
Sorry, I didn't fully phrase what I meant.
Something like this:
```dart
if (await _targetFile!.length() > maxLogFileSizeMB * 1000000) {
  try {
    await _closeCurrentFile();
  } catch (e, s) {
    print(e);
    print(s);
  } finally {
    await _openFile(proposed);
  }
}
```
_I also added logging, otherwise the developer has little chance to troubleshoot the problem (should one arise)._

I was also under the initial assumption that `_closeCurrentFile` was the only real "catch-able" exception (sensible to catch and continue) in this block.
However, I am not sure if `sink.flush()` or `sink.close()` ever really throws, and it would probably make more sense to also guard the `length` check (as you did).

To sum it up, I would just like you to add the logging and guard the second call to `_closeCurrentFile` (as we can actually continue without successfully closing the old file but not without a new file).

Reply:
Okay, I'll add logging. About catchable exceptions, _targetFile!.length() can throw if the target file is missing. This can happen if an external user or program deleted it while the app was running. On newer versions of Dart, it throws somethings like FileMissingException (forgot the actual name), but it's not yet available on Dart 2.17, so I replaced it with a general catch clause for all exceptions.

Reply:
Thanks!
Exactly, yeah, don't restrict the type, I don't think any exceptions should be propagated here.